### ISSUE TITLE
Slow moving collision objects get stuck on collision (DEF-2488)

### ISSUE THREAD

**Issue by britzl** [defold/defold#3775]
Description:

If a dynamic collision object has a very low velocity and collides with another object it is likely to get stuck to the object while still retaining some velocity and move along the collision shape. This is caused by a Box2D configurable valueb2_velocityThreshold that by default is set to 1.0f. We should expose this value in game.project.

Steps to reproduce:

Let a dynamic collision object move slowly and collide with another collision object

Expected result:

If Box2D is configured properly the colliding object should still bounce

Actual result:

The colliding object sticks to the other collision object while retaining speed and following the shape.

Reported on the forum:(LINK REMOVED)

**Comment by otsakir** [comment]
hello @britzl !

I reproduced the behavior with one moving object bumping onto another. If the velocity of the first is <= 1.0 , the collision is inelastic and they move as one afterwards. 

I checked bundled Box2D source code for `b2_velocityThreshold`. Here is what i got:

```
$ grep -r b2_velocityThreshold *
engine/physics/src/box2d/Box2D/Common/b2Settings.h:#define b2_velocityThreshold         1.0f
engine/physics/src/box2d/Box2D/Dynamics/Contacts/b2ContactSolver.cpp:                   if (vRel < -b2_velocityThreshold)
```

So, as expected it is set to 1.0f in the code. The downside is that the way this is set and used is not flexible (a C preprocessor definition). 

No clean solution comes to mind. Maybe set up some sort of interface between box2d and the engine so that there is a function/variable provided by the engine that returns the value defined in `game.project` as `velocityThreshold`. Then box2d bundled code should be  modified so that it looks that up instead of using the #defined one. 

Thoughts ?

**Comment by JCash** [comment]
We have a few places where we do use Defold specific code in Box2D/Bullet3D.
This would be another such place.

Make it an actual function+variable
E.g. a static function `b2ContactSolver::setVelocityThreshold(float threshold)`
and we can set it in `physics_2d.cpp` using the `NewWorldParams params`.

**Comment by otsakir** [comment]
@JCash , i noticed that when creating a new World2D there are two pieces that contain initialization information. A `Context2D` and `NewWorldParams`. From my understanding, the new 'velocityThreshold' property will be part of the latter. The reason here is that the property will be used to set up the physics world _only_ and is not really required throughout the game. 

https://github.com/defold/defold/blob/dev/engine/physics/src/physics/physics_2d.cpp#L247

Am i guessing right ?

**Comment by otsakir** [comment]
I'm thinking that the proper place to call `b2ContactSolver::setVelocityThreshold()`  is  in NewContext2D(). Seems a better fit than NewWorld2D() since it's initialized only once. 

https://github.com/defold/defold/blob/dev/engine/physics/src/physics/physics_2d.cpp#L218

**Comment by otsakir** [comment]
Got a working version of the feature. The new `velocityThreshold` property is not present in the editor but it works if edited directly in game.projectc.

A minor concern is that the point when the threshold gets effective seems to depend on the physics `scale` property. So with scale=1, a `velocityThreshold` as small as 0.9 is enough to make the object bounce off one another. Instead, if scale=0.02,  a lower velocityThreshold  like 0.1 is needed. 0.5 for instance will still produce an inelastic collision.

**Comment by otsakir** [comment]
@JCash , wrt to adding the `velocityThreshold` to the editor, is there anything to do besides putting the property here:

https://github.com/defold/defold/blob/dev/editor/resources/meta.edn

?

**Comment by JCash** [comment]
Yes, there's also `meta.properties`for bob.jar:
https://github.com/defold/defold/blob/dev/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties

**Comment by britzl** [comment]
> A minor concern is that the point when the threshold gets effective seems to depend on the physics `scale` property. So with scale=1, a `velocityThreshold` as small as 0.9 is enough to make the object bounce off one another. Instead, if scale=0.02, a lower velocityThreshold like 0.1 is needed. 0.5 for instance will still produce an inelastic collision.

Does this mean that we should take the physics scale in consideration at runtime before using `velocityThreshold` or should we leave that to the developer when setting the value in game.project?

**Comment by otsakir** [comment]
@britzl , it should be possible to have velocityThreshold calculated so that it takes into account the `scale` property. At least mathematically there is a pattern. 

So, suppose that one has a scale property of 0.02, and a marginal velocity of 1000. So, if objects run with velocity under 1000 they shoud stick together. The respective `velocityThreshold` property given to Box2D should be 1000*0.02 = **50**. Maybe subtract a little bit to be on the safe side. 

It's an easy rule to remember but i guess we could also implement that in our code too.

**Comment by JCash** [comment]
If we use the scaling in the code to make these conversions, we should use it for this too?
I think consistency is good.

Although, this property comes from the user settings, and we _could_ tell them to keep those two values in sync, but that might be a bit unwieldy if they need to tweak things (i.e. when changing scale)

I think it might be better for us to do the conversion at runtime, and in the documentation for `velocityThreshold` we let them know that the value is modified by the scale in the engine.

**Comment by otsakir** [comment]
Great. Let me implement that in the code then.

**Comment by otsakir** [comment]
@JCash , about testing i'm working on gamesys module scenario where two dynamic object's are standing stand on a static base. I set the velocity of the first and after the collision i check the resulting velocities of both objects to determine if they move as one or not. 

Anything else that comes to mind ?

**Comment by JCash** [comment]
Just a reminder that we reverted this commit since it was contributing to our branch failing.
da10b2415b5dceab746a1e0478e0c63897dc0623

**Comment by otsakir** [comment]
Reportedly, the `ScaledImpulses` test fails.

@JCash , any hint about the context this test fails in ?

**Comment by otsakir** [comment]
I went through `ScaledImpulses` test. By reading it that is. 

This is a strictly sequential test. An object falls and after 400 step iterations its Y position is verified by a high accuracy comparison:

```
ASSERT_NEAR(y, vo_a.m_Position.getY(), 0.00001f);
```

This is a very accurate operation. My first guess is that there is some sort of rounding that happens when i override the velocity threshold that produces a slightly different value and causes the failure.

**Comment by otsakir** [comment]
Note, in the specific test the context 'Scale' attribute is explicitly set to 0.5f test_physics.h:

```
static const float PHYSICS_SCALE = 0.5f;

template<typename T>
class PhysicsTest : public jc_test_base_class
{
protected:

    virtual void SetUp()
    {
        dmPhysics::NewContextParams context_params = dmPhysics::NewContextParams();
        context_params.m_Scale = PHYSICS_SCALE;
```

https://github.com/defold/defold/blob/dev/engine/physics/src/physics/test/test_physics.h#L40-L50

**Comment by JCash** [comment]
Not sure why it fails tbh. What is the error value? I.e. the difference between the two y positions?

**Comment by otsakir** [comment]
Well, like i said, i couldn't reproduce. The sole assertion in the test is this:

ASSERT_NEAR(y, vo_a.m_Position.getY(), 0.00001f);

But meanwhile, more evidence came up :-). The following is the outcome of running ScaledImpulses test two times in a row. Both in my linux machine:

```
10:12:28 ~/src/defold/engine/physics (dev_3775) $ ./build/default/src/physics/test/test_physics --test-filter ScaledImpulses
PhysicsTest
ScaledImpulses/0 
ScaledImpulses/0  PASS (2298 µs)
PhysicsTest took 2434 µs
PhysicsTest
ScaledImpulses/1 
ScaledImpulses/1  PASS (4506 µs)
PhysicsTest took 7.44 ms
Ran 2 tests, with 20 assertions in 9.874 ms
2 tests PASSED and 82 skipped
```
and the second
```
10:12:29 ~/src/defold/engine/physics (dev_3775) $ ./build/default/src/physics/test/test_physics --test-filter ScaledImpulses
PhysicsTest
ScaledImpulses/0 
../src/physics/test/test_physics.cpp:1827:
Value of: abs(y - vo_a.m_Position.getY()) <= 0.00001f
Expected: abs(1.001999 - 1.003599) <= 0.000010
  Actual: abs(1.001999 - 1.003599) == 0.001600

ScaledImpulses/0  FAIL (2054 µs)
PhysicsTest took 2138 µs
PhysicsTest
ScaledImpulses/1 FATAL:PHYSICS: Could not create socket '@physics'.
Segmentation fault (core dumped)
```

So, the test is not 100% deterministic. And since it depends on accurate calculations it sometimes fails. I ran it 11 times. It failed on 2.

**Comment by otsakir** [comment]
Do you by any change remember the error you got ? Is it the same ?

**Comment by otsakir** [comment]
Finally got to the bottom of it.

`m_VelocityThreshold`, though properly initialized in test_gamesys.h (gamesys module),  it's not initialized in test_physics.h (physics module). Thus, the `physics` tests may be affected by it. `ScaledImpulse` unit test is sensitive by nature and sporadically failed.

**Comment by otsakir** [comment]
@JCash , ready to go. Shall i create a new PR ?

**Comment by AGulev** [comment]
## Short issue summary

The report says slow-moving dynamic 2D collision objects can stick to other objects instead of bouncing because Box2D uses a fixed restitution/velocity threshold. The issue discussion proposed exposing that threshold through `game.project` so developers can tune it.

## Status

Possibly outdated/fixed

## Evidence

- The GitHub issue is still `OPEN`, but the discussion shows the work was implemented and merged through PR `#6076` and a follow-up fix PR `#6205`, both merged in 2021.
- The engine now reads `physics.velocity_threshold` from config in `engine/engine/src/engine.cpp:1364`.
- Physics context creation stores that value by default in `engine/physics/src/physics/physics_common.cpp:24` and passes it into the 2D physics backends in `engine/physics/src/physics/box2d_defold/box2d_defold_physics.cpp:222` and `engine/physics/src/physics/box2d/box2d_physics.cpp:158`.
- The legacy Box2D integration added an override hook for the old fixed threshold in `external/box2d_v2/Box2D/Dynamics/Contacts/b2ContactSolver.cpp:30` and uses it when deciding whether restitution should apply in `external/box2d_v2/Box2D/Dynamics/Contacts/b2ContactSolver.cpp:215`.
- The newer Box2D path sets `worldDef.restitutionThreshold` from the configurable value in `engine/physics/src/physics/box2d/box2d_physics.cpp:65`.
- Build metadata includes the project setting in `com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties:226`.
- There is a dedicated regression test for this behavior in `engine/gamesys/src/gamesys/test/test_gamesys.cpp:3092`, with custom test setup in `engine/gamesys/src/gamesys/test/test_gamesys.h:227` and scripted assertions in `engine/gamesys/src/gamesys/test/collision_object/velocity_threshold.script:15`.

## Why this looks already fixed

The original problem was that the velocity threshold was effectively fixed inside Box2D, so low-speed contacts could become inelastic with no project-level override. Current Defold code now exposes `physics.velocity_threshold`, wires it into both 2D physics implementations, and includes a regression test specifically for the slow-collision behavior discussed in the issue. That makes the current open issue look stale rather than still unresolved in the codebase.
